presto-kinesis/src/main/java/io/prestosql/plugin/kinesis/s3config/S3TableConfigClient.java

```diff
@@ -0,0 +1,266 @@
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.AmazonS3URI;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.ListObjectsRequest;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Inject;
+import io.airlift.json.JsonCodec;
+import io.airlift.log.Logger;
+import io.prestosql.plugin.kinesis.ConnectorShutdown;
+import io.prestosql.plugin.kinesis.KinesisClientProvider;
+import io.prestosql.plugin.kinesis.KinesisConnectorConfig;
+import io.prestosql.plugin.kinesis.KinesisStreamDescription;
+import io.prestosql.spi.connector.SchemaTableName;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Utility class to retrieve table definitions from a common place on Amazon S3.
+ * <p>
+ * This is so that we can add new tables in a central "metastore" location without
+ * having to update every single node with the files.
+ * <p>
+ * This makes calls to Amazon AWS using the S3 client.
+ */
+public class S3TableConfigClient
+        implements ConnectorShutdown
+{
+    private static final Logger log = Logger.get(S3TableConfigClient.class);
+
+    public final KinesisConnectorConfig kinesisConnectorConfig;
+    private final KinesisClientProvider clientManager;
+    private final JsonCodec<KinesisStreamDescription> streamDescriptionCodec;
+
+    private final String bucketUrl;
+    private long lastCheck;
+    private ScheduledFuture<?> updateTaskHandle;
+
+    private Map<String, KinesisStreamDescription> internalMap =
+            Collections.synchronizedMap(new HashMap<String, KinesisStreamDescription>());
+
+    @Inject
+    public S3TableConfigClient(KinesisConnectorConfig aConnectorConfig,
+            KinesisClientProvider aClientManager,
+            JsonCodec<KinesisStreamDescription> jsonCodec)
+    {
+        this.kinesisConnectorConfig = requireNonNull(aConnectorConfig, "connector configuration object is null");
+        this.clientManager = requireNonNull(aClientManager, "client manager object is null");
+        this.streamDescriptionCodec = requireNonNull(jsonCodec, "JSON codec object is null");
+
+        // If using S3 start thread that periodically looks for updates
+        this.bucketUrl = this.kinesisConnectorConfig.getTableDescriptionsS3();
+        if (!this.bucketUrl.isEmpty()) {
+            startS3Updates();
```

Comment:
This is not safe from a Java memory model perspective, since the thread can access this object before it is fully constructed. Move the thread creation to a method annotated with `@PostConstruct` -- it will be called after the object is created.

Reply:
moved to post construct